src/collections.h

```diff
@@ -313,6 +369,83 @@ Collection<T>::fold(std::function<T(T, T)> func) {
 
 
 template<typename T>
+T
+Collection<T>::reduceRight(std::function<T(T, T)> func) {
+    // TODO: bounds checking
+    T val = func(Data[Data.size() - 1], Data[Data.size() - 2]);
+    for (int i = Data.size() - 3; i >= 0; i--)
+        val = func(val, Data[i]);
+
+    return val;
+};
+
+
+template<typename T>
+void *
+preduce_thread(void *arg) {
+    struct thread_data {
+        std::vector<T> *list;
+        std::function<T(T, T)> func;
+        int begin;
+        int end;
+        T retval;
+    };
+    thread_data *data = ((struct thread_data *)arg);
+    std::vector<T> list = *(data->list);
+
+    T val = (data->func)(list[data->begin], list[data->begin + 1]);
+    for (int i = data->begin + 2; i < data->end && i < list.size(); i++)
+        val = (data->func)(val, list[i]);
+    data->retval = val;
+    pthread_exit(NULL);
+};
+
+
+template<typename T>
+T
+Collection<T>::preduce(std::function<T(T, T)> func, int threads) {
+    struct thread_data {
+        std::vector<T> *list;
+        std::function<T(T, T)> func;
+        int begin;
+        int end;
+        T retval;
+    };
+    std::vector<pthread_t> thread_pool(threads);
+    std::vector<thread_data> thread_data_pool;
+
+    int chunk = Data.size() / threads;
+    int extra = Data.size() - chunk*threads;
+    std::vector<int> indices(extra, chunk+1);
+    std::vector<int> normal(threads-extra, chunk);
+    indices.insert(indices.end(), normal.begin(), normal.end());
+
+    int start = 0;
+    for (int i = 0; i < threads; i++) {
+        pthread_t pid;
+        thread_pool[i] = pid;
+
+        int end = start + indices[i];
+        thread_data td = { &Data, func, start, end};
+        thread_data_pool.push_back(td);
+        start = end;
+    }
+
+    for (int i = 0; i < threads; i++)
+        pthread_create(&(thread_pool[i]), NULL, preduce_thread<T>, &(thread_data_pool[i]));
+
+    for (pthread_t i : thread_pool)
+        pthread_join(i, NULL);
+
+    T val = func(thread_data_pool[0].retval, thread_data_pool[1].retval);
```

Comment:
we need bounds checking, here. wanna add a not so we do it later w/ every other method?